framework-plugins/lisk-framework-chain-connector-plugin/src/chain_connector_plugin.ts

```diff
@@ -677,7 +686,7 @@ export class ChainConnectorPlugin extends BasePlugin<ChainConnectorPluginConfig>
 		const blockHeaders = await this._chainConnectorStore.getBlockHeaders();
 
 		const updatedBlockHeaders = blockHeaders.filter(
-			blockHeader => blockHeader.height >= this._lastCertificate.height,
+			blockHeader => blockHeader.height >= this._receivingChainFinalizedHeight,
```

Comment:
Since the sending chain and receiving chains are different they will have different block heights so the above cleanup will behave randomly. 

So taking an example,
- Sending chain(S1) current block height === 50
- Receiving chain(R1) current block height === 180
- lastCertificate height === 41 of S1 chain on receiving chain at height 175
- finalizedHeight of R1 === 171

So in the above case we only delete chain connector data when finalized height on R1 >= 175 where the lastcertificate height===41 of S1 was included.

one solution could be, 
we declare,
```ts
private _receivingChainInfo!: { currentHeight: number; finalizedHeight: number; lastCertificateHeightOfSendingChain: number; lastCertificateBlockHeightOnReceivingChain: number }
```

On every new block we update, `_receivingChainInfo.finalizedHeight` and `_receivingChainInfo.currentHeight` by calling `getNodeInfo` on receiving chain.

On cleanup, we check, 

- If `_receivingChainInfo.finalizedHeight >= lastCertificateBlockHeightOnReceivingChain`
  - We delete all the records < `_receivingChainInfo.lastCertificateHeightOfSendingChain`
  - Update `_receivingChainInfo.lastCertificateHeightOfSendingChain = this.lastCertificate.height`
  - Update `_receivingChainInfo.lastCertificateBlockHeightOnReceivingChain = _receivingChainInfo.currentHeight`
   
- Else, do anything

Reply:
We will keep in memory `heightToDeleteIndex` which stores finalized height to inbox size and last certificate height and use this information during the cleanup process